bot/cogs/moderation/infractions.py

```diff
@@ -0,0 +1,563 @@
+                Member: {user.mention} (`{user.id}`)
+                Actor: {ctx.message.author}{dm_log_text}
+                Reason: {reason}
+                {expiry_log_text}
+            """),
+            content=log_content,
+            footer=f"ID {infraction['id']}"
+        )
+
+    async def pardon_infraction(self, ctx: Context, infr_type: str, user: MemberObject) -> None:
+        """Prematurely end an infraction for a user and log the action in the mod log."""
+        # Check the current active infraction
+        response = await self.bot.api_client.get(
+            'bot/infractions',
+            params={
+                'active': 'true',
+                'type': infr_type,
+                'user__id': user.id
+            }
+        )
+
+        if not response:
+            await ctx.send(f":x: There's no active {infr_type} infraction for user {user.mention}.")
+            return
+
+        # Deactivate the infraction and cancel its scheduled expiration task.
+        log_text = await self.deactivate_infraction(response[0], send_log=False)
+
+        log_text["Member"] = f"{user.mention}(`{user.id}`)"
+        log_text["Actor"] = str(ctx.message.author)
+        log_content = None
+        footer = f"ID: {response[0]['id']}"
+
+        # If multiple active infractions were found, mark them as inactive in the database
+        # and cancel their expiration tasks.
+        if len(response) > 1:
+            log.warning(f"Found more than one active {infr_type} infraction for user {user.id}")
+
+            footer = f"Infraction IDs: {', '.join(str(infr['id']) for infr in response)}"
+
+            log_note = f"Found multiple **active** {infr_type} infractions in the database."
+            if "Note" in log_text:
+                log_text["Note"] = f" {log_note}"
+            else:
+                log_text["Note"] = log_note
+
+            # deactivate_infraction() is not called again because:
+            #     1. Discord cannot store multiple active bans or assign multiples of the same role
+            #     2. It would send a pardon DM for each active infraction, which is redundant
+            for infraction in response[1:]:
+                _id = infraction['id']
+                try:
+                    # Mark infraction as inactive in the database.
+                    await self.bot.api_client.patch(
+                        f"bot/infractions/{_id}",
+                        json={"active": False}
+                    )
+                except ResponseCodeError:
+                    log.exception(f"Failed to deactivate infraction #{_id} ({infr_type})")
+                    # This is simpler and cleaner than trying to concatenate all the errors.
+                    log_text["Failure"] = "See bot's logs for details."
+
+                # Cancel pending expiration task.
+                if infraction["expires_at"] is not None:
+                    self.cancel_task(infraction["id"])
```

Comment:
This code handles multiple active infractions. Does it still need to be here for the time being or can it already be removed given what you are working on? I'm not caught up on the status of that. @SebastiaanZ

Reply:
Double active infractions may still occur for now. The code for the PR is done, the migrations file as well, but it will be stalled until python-discord/site#269 is resolved, since that PR also adds a migration file (and is actually far more critical). (We can't merge them independently because migration files have a single, explicit parent file mentioned in their body.)

I was debating with myself not to include the data migration and open a PR later, to prevent the conflict and not stall the validation, but that would still leave users with multiple active infractions in the database (17 users have double bans; no users have double mutes), it only guarantees we won't add more using the API.

Reply:
Thanks. Maybe everything will get merged in time but that seems unlikely. I'll leave this unresolved so I don't forget.